### ISSUE TITLE
Allow per-document notebook content options

### ISSUE THREAD

**Issue by firelizzard18** [microsoft/vscode#120600]
I want to be able to set `transientOutputs` on a per-notebook document basis.

For example, say I invent some notebook format that provides a way to store per-document configuration values. One of those values controls whether or not outputs are persisted. In this scenario, some notebooks should have `transientOutputs: true` and others `transientOutputs: false`. To reflect this properly, I want to be able to set `NotebookDocumentContentOptions` on a per-notebook basis, probably in the content provider.

As for why that would be useful: If I am committing notebooks to a repository, I most likely do not want to persist outputs, especially if they are dependent on time/machine/location/etc (i.e. if they are liable to change). However, if I have a notebook outside of source control (that I'm using for tests/experiments/etc), I likely want outputs to persist.

**Comment by rebornix** [comment]
I wonder if this can be an extension feature and extension can decide if they want outputs to be stored on disk per document or cell. Thought the catch is the dirty indicator is not reflecting the state.

**Comment by firelizzard18** [comment]
@rebornix I was thinking of this as a capability that an extension can choose to use or not - something like an optional field on `NotebookData` that overrides the content provider's options if the field is present. So the extension's content provider controls this.

I can do this today, but as far as I can tell, the entire purpose of the transient flags is to make sure the dirty indicator is updated correctly. And without being able to set this per-notebook, I have to choose between false positives or false negatives for the dirty indicator.

For a more concrete example, I'm imagining a JSON notebook format corresponding to the following:

```typescript
type Notebook = {
    options: NotebookOptions;
    cells: NotebookCells[];
}

type NotebookOptions = {
    saveOutputs: boolean;
}

type NotebookCells = {
    kind: 'markdown' | 'code';
    language: string;
    content: string;
    outputs: any;
}
```

It would be the responsibility of the extension author to provide a way to change the flag and communicate it to the user.

**Comment by jrieken** [comment]
Yeah, I was having similar thoughts when persisting output for GitHub issue notebooks. Tho, for that it would have been per output/metadata value. E.g when a cell contains `@me` I don't want to save its output and that fact is very dynamic (changes after providing `NotebookData`)

**Comment by DonJayamanne** [comment]
I too have a similar requirement for kusto notebooks.

When looking at some telemetry i'd like to save the graphs & the output
E.g. a pie chart that represents how many failures we had. this information doesn't contain any PII.

I'd like to keep this information for historical purposes. so when i run this again next week as a new cell, i can compare it against the previous week (another cell).

The only way I can think of doing this is;
* Create cell A and run it, save output
* next week, create cell B with same code as cell A, run it 
* now compare output of A & B

**Comment by jrieken** [comment]
One idea is to differentiate between metadata and output, so for metadata we declare per provider (maybe per document) what is transient and what is persistent but for outputs we make it an instance properly. E.g in its simplest form a we can have a `transient?: boolean`-property in `NotebookOutput` and that would express "this output is not going to be saved"

**Comment by firelizzard18** [comment]
That seems reasonable to me. If for some reason I did want some cells to have transient metadata and others to have persistent metadata, I could define separate custom metadata keys, one persisted, one transient. Not that I have a reason to do that at the moment.

**Comment by jrieken** [comment]
>  I could define separate custom metadata keys, one persisted, one transient. Not that I have a reason to do that at the moment.

Yes - that exactly was our reasoning as well. But once transient it is so for all cells

**Comment by jrieken** [comment]
blocked by https://github.com/microsoft/vscode/issues/121807

**Comment by sourishkrout** [comment]
> blocked by #121807

@jrieken + @rebornix: it looks like this is no longer blocked by the above issue (cc @firelizzard18).

I second the original feature request. It also seems odd that once `transientOutputs: false`, cells are passed into the serializer without their respective outputs. The `transientOutputs` option in the context of dirty-tracking in the UI the option makes absolute sense. However, should it not be up to the notebook's serializer implementation to decide what to do with outputs in either case? In my particular implementation outputs are indeed considered transient, however, that does not mean they are exempt from serialization.

I can file a separate issue if it's better considered independently.
